File: backend/app/models/project.py
from pydantic import BaseModel, HttpUrl
from typing import List, Optional
from datetime import datetime

# Request Models
class ProjectCreate(BaseModel):
    name: str
    start_date: str
    end_date: str  # Can be "Present" or date
    tech_stack: str
    link: Optional[str] = None
    link_label: Optional[str] = None  # e.g., "GitHub", "GitLab", "Live Demo", etc.
    subpoints: List[str] = []

class ProjectUpdate(BaseModel):
    name: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    tech_stack: Optional[str] = None
    link: Optional[str] = None
    link_label: Optional[str] = None
    subpoints: Optional[List[str]] = None

# Response Models
class ProjectResponse(BaseModel):
    id: str
    user_id: str
    name: str
    start_date: str
    end_date: str
    tech_stack: str
    link: Optional[str] = None
    link_label: Optional[str] = None
    subpoints: List[str] = []
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

File: frontend/src/api/project.ts
import api from './axios'

export interface ProjectResponse {
  id: number
  user_id: number
  name: string
  start_date: string
  end_date: string
  tech_stack: string
  link: string | null
  link_label: string | null
  subpoints: string[]
  created_at: string
  updated_at: string
}

export interface ProjectCreate {
  name: string
  start_date: string
  end_date: string
  tech_stack: string
  link?: string | null
  link_label?: string | null
  subpoints?: string[]
}

export interface ProjectUpdate {
  name?: string
  start_date?: string
  end_date?: string
  tech_stack?: string
  link?: string | null
  link_label?: string | null
  subpoints?: string[]
}

export const projectApi = {
  getProject: async (): Promise<ProjectResponse[]> => {
    const response = await api.get<ProjectResponse[]>('/api/project/')
    return response.data
  },

  getProjectById: async (projectId: number): Promise<ProjectResponse> => {
    const response = await api.get<ProjectResponse>(`/api/project/${projectId}`)
    return response.data
  },

  createProject: async (data: ProjectCreate): Promise<ProjectResponse> => {
    const response = await api.post<ProjectResponse>('/api/project/', data)
    return response.data
  },

  updateProject: async (projectId: number, data: ProjectUpdate): Promise<ProjectResponse> => {
    const response = await api.put<ProjectResponse>(`/api/project/${projectId}`, data)
    return response.data
  },

  deleteProject: async (projectId: number): Promise<void> => {
    await api.delete(`/api/project/${projectId}`)
  },
}
